### ISSUE TITLE
Multiple Client ID per Application

### ISSUE THREAD

**Issue by ICEMANno1** [mitreid-connect/OpenID-Connect-Java-Spring-Server#1119]
Hi, is it planned to have multiple Client IDs per application? 
Use Case could be an Mobile App which wants to substitute it's client I'd on a yearly base. To support a transition phase over some time, the developer wants to have two client IDs for one app. Therefore he can monitor the usage of the "old" client ID and if there is almost no usage anymore decide to delete the old ID and only support the new one.
Thanks

**Comment by jricher** [comment]
We have no plans to support multiple client IDs for a given client application. Why would you rotate a client ID on an annual basis? What's the use case in that?

**Comment by ICEMANno1** [comment]
Main reason: reducing the attack vector. Same reason why you rotate certificates of website.

**Comment by jricher** [comment]
But the client ID isn't a secret and isn't treated as one by the underlying protocols. As soon as you use it, it's known to the user and the browser and could at that instant be copied to anything else. That's why you use client secrets (which can be rotated) or PKCE (for public clients with no secret).

**Comment by ICEMANno1** [comment]
I agree that the client ID is not a secret for certain flows (e.g. implicit flow). Client secrets don't add any value on a public client for the same reasons like for the client ID. As far as I could see in the code and DB, each app can only have one secret as well.
Nevertheless, in Oauth2 RFC6749 it states in section 3.2.1: "Implementing authentication management best practices, which require periodic credential rotation.  Rotation of an entire set of refresh tokens can be challenging, while rotation of a single set of client credentials is significantly easier." 
Therefore I think the writers of it had the same thoughts as I had, or?

**Comment by jricher** [comment]
The authors did not have the same thoughts as you. We intended that section to refer to rotation of the client secret and other credentials such as tokens, not the client ID. The client ID is an identifier, as the name suggests, not a credential.

**Comment by ICEMANno1** [comment]
OK. How would the rotation of the client secret work?

**Comment by jricher** [comment]
On this server package, it depends on the type of registration. For static registration, just open up the client's registration page and select the "generate client secret" option. For dynamic registration, you'd use the GET operation on the management endpoint, but we haven't implemented automatic rotation there yet (see #402).

**Comment by ICEMANno1** [comment]
Thanks for the fast reply. What if you have multiple server instances and want to do a gradual role over for those to avoid a downtime? Would 2 to n secrets with e.g. an expiry help?

**Comment by jricher** [comment]
If you have multiple authorization servers, you should NOT be sharing the client ID and secret among them. It's considered best practice to use a different ID and secret at every authorization server. If you have multiple faces of an authorization server (for load balancing or something like that) then you'd be better served by synchronizing your storage layer, since you'll need to deal with token storage as well.

**Comment by ICEMANno1** [comment]
Sorry, I didn't state the use case clear enough. I was talking about multiple oauth2 client servers e.g. webserver or I think oauth called them confidential web application.

**Comment by jricher** [comment]
If you have multiple clients then they should each definitely have their own ID and secret, or else use some other kind of data synchronization to act as a single application.

**Comment by ICEMANno1** [comment]
Well, they are one application with multiple instances. If multiple Client Id and client secret pairs would be possible per application that would be the best solution for both: credential rotation and multiple instances of one application. Then there would not be an additional effort to sync client credentials and enable more enterprise like use cases.

**Comment by jricher** [comment]
In that case, each instance usually gets its own ID and secret, and they shouldn't be shared.

**Comment by ICEMANno1** [comment]
But why did you bring in the level of application in your model then, if anything related needs to have it's own app instance incl. Client credentials etc.?

**Comment by jricher** [comment]
We didn't. Each client has its own unique client ID (this is enforced), and that's how the model is built in this software.

**Comment by ICEMANno1** [comment]
Well, I confused myself here :) sorry! What I meant is the meta data which is stored for each client, e.g. description, logo, ToS, homepage contacts, etc. needs to be maintained several times if it's one application. So wouldn't it make sense to get the application level in, which has common information about several clients sorted only once? Then all security can stay intact e.g. you just create a new client if you want to rotate anything and delete the old when done.

Thoughts on this?

**Comment by ICEMANno1** [comment]
P.s. thanks for your effort with me ;)

**Comment by jricher** [comment]
That's not the model that we have here. The unique index is the client ID, and if you've got multiple client IDs (and multiple instances of the same software) then each is managed separately. I suppose you could build it with another layer of abstraction, but we haven't done that, and doing so would require significant rewrites of the application from a deeply fundamental level.

**Comment by ICEMANno1** [comment]
Well, so I guess you won't plan to add layers like Google has: Project with multiple members and multiple Clients + cross client auth?

https://developers.google.com/identity/protocols/CrossClientAuth

**Comment by jricher** [comment]
Such features are not on our roadmap at this time. That said, since development on 1.3 was recently opened, data models can change and if there's significant interest in it (especially if someone's willing to put in a pull request) then this isn't out of the realm of possibility. But it's not something we're pursuing right now.

**Comment by ICEMANno1** [comment]
How would you usually handle this? Should I start a new issue as feature request, describe it a bit more and you discuss onwhich changes are required and how they should be done?

**Comment by jricher** [comment]
You can open a new issue if you want, but it's not something we're planning on building right now. The best way to get it in would be to have a pull request, which we can then eventually review and possibly incorporate.

**Comment by ICEMANno1** [comment]
Like you said it's a rather big change. Therefore I guess it my be better to discuss first before creating a pull request which changes way to much at the wrong places.
The idea was 1st describe the solution and agree on it, 2nd discuss the places to do the changes and agree on it and 3rd create a pull request to do the changes and then review them.
Your thoughts?
